Add vision server and drive logic

WasabiFan opened the pull request:
This adds the vision server and drive logic. The changes that will immediately affect the robot at runtime are:
- A server is started on UDP port 5800. This code is surrounded in error handling so it shouldn't be able to cause fatal errors, and starting a server requires negligible resources; a listener is registered and a new thread sleeps waiting for the `receive()` call to return.
- Four testing commands are added to left joystick buttons 7, 8, 9 and 10.

The latter half of this diff adds the source of a JSON parsing library (so that no additional JARs have to be loaded).

WasabiFan committed:
Add initial vision logic

WasabiFan committed:
Fix everything

WasabiFan committed:
Fix all the things (take two)

Commit:
merge with master

WasabiFan committed:
Merge branch master

WasabiFan committed:
Update vision code for recent changes on "master"

Commit:
Merge remote-tracking branch 'origin/master' into vision

# Conflicts:
#	SeriouslyCommonLib

Commit:
Robot To Boiler Command Group (#59)

* fix build error

* point to private imask

* Move to hopper command group

* Move To Hopper Command with Rotate Robot to boiler command

* Reference version of common lib with checkstyle suppressions

* Added ShooterBelts to Smart Dashboard. (#62)

* Added ShooterBelts to Smart Dashboard.

* Changed "ShooterBelt" to "ShooterBelts"

* Added ShooterBelt intake and eject power log.

* Feature/shoot fuel command (#60)

* Unfinished changes (2/16/2017).

* Created high-level "Shoot Fuel" command.

* Fixed Travis build errors.

* Fixed Travis Build errors.

* Created "ShootFuelCommandGroup" that runs both left and right fuel command groups.

* Made changes requested from Pull Request.

* Made changes requested from Pull Request. (reverted from commit 85616c905a746efa00b94c374ca8e717f661f587)

* Made changes from Pull Request.

* Feature/autonomous command selection (#61)

Created a autonomous command selector. Mapped the base autonomous command group/code to be activated by Smart DashBoard

* Deal with XbotButton(s) build break

* Initial Oi Testing Outline

Mapped out all commands needed to actuate robot to oi or smart dashboard.

* Detailed End logging for Drive Commands

Created end logging for drive commands

* Basically A LOT of chanages made :)

* modify CalculateDistanceOfParallelCommand.java

* point to newer commonlib

* fix merge issues

WasabiFan committed:
Misc vision updates and fixes

WasabiFan committed:
Merge branch master into vision

WasabiFan committed:
Use "parallel to boiler" angle for heading command

WasabiFan committed:
Merge master

WasabiFan committed:
Update common lib

WasabiFan committed:
Merge remote-tracking branch 'origin/master' into vision

WasabiFan committed:
Fix minor bugs in vision drive code

Commit:
Implement better logging (#75)

* possibly implemented better logging.

* did stuff

* Some more requested changes made

* Fixed build errors

* faffjdjs

* add checks to confirm properties exist

* merge changes undone

* Fix bad merge

* More merge fixes

WasabiFan committed:
Merge branch 'vision' of https://github.com/team488/TeamXbot2017 into vision

WasabiFan committed:
Clean up vision subsystem logic

WasabiFan committed:
Misc minor clean-up of vision code

WasabiFan committed:
Merge branch master into vision

WasabiFan committed:
Update common lib

WasabiFan committed:
Merge remote-tracking branch 'origin/master' into vision

imathno commented on Competition2017/src/competition/networking/NetworkedCommunicationServer.java:
```diff
@@ -0,0 +1,153 @@
+
+        public void stopServer() {
+            this.isRunning = false;
+            serverSocket.close();
```
Should we log when the server is stopping?

WasabiFan committed:
Fix vision drive and logging bugs

WasabiFan committed:
Use injected comm server

WasabiFan committed:
Add mock implementaton of comm server

WasabiFan committed:
Merge remote-tracking branch 'origin/master' into vision

WasabiFan committed:
Properly bind vision server in test env

JohnGilb commented on Competition2017/src/competition/operator_interface/OperatorCommandMap.java:
```diff
@@ -170,6 +173,24 @@ public void setupDriveCommand(
+            DriveToBoilerWithVisionCommandGroup driveToBoilerCommand,
+            SetRobotHeadingCommand setHeadingCommand,
+            RotateToHeadingCommand rotateToHeadingCommand
+    )   {
```
At this point, we've moved all debug functions to the dashboard instead of the joystick, meaning that the upper joystick buttons (the easier ones to press) are available. I suggest putting your commands up there now, to increase the chance that the drivers will try them out and acclimate to them.

JohnGilb commented on Competition2017/src/competition/operator_interface/OperatorCommandMap.java:
```diff
@@ -170,6 +173,24 @@ public void setupDriveCommand(
+        oi.leftButtons.getifAvailable(7).whileHeld(rotateCommand);
+        oi.leftButtons.getifAvailable(8).whileHeld(driveToBoilerCommand);
+        
+        setHeadingCommand.setHeadingToApply(248);
```
Is there some significance to this number?

JohnGilb replied:
I see it's "Angle of line parallel to boiler". I'd suggest putting that as a static variable in the pose subsystem or the vision subsystem.

JohnGilb commented on Competition2017/src/competition/subsystems/drive/commands/CalculateDistanceOfBoilerParallelCommand.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Override
+    public void initialize() {
```
Most of our commands include an "initializing" log to help us track runtime flow. I'd recommend it here and in other new commands.

JohnGilb commented on Competition2017/src/competition/subsystems/drive/commands/DriveInfinitelyCommand.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public void initialize() {
+        power = powerTargetSupplier == null ? 0 : powerTargetSupplier.getAsDouble();
```
same "Initializing"

JohnGilb commented on Competition2017/src/competition/subsystems/vision/VisionSubsystem.java:
```diff
@@ -0,0 +1,165 @@
+
+    @Inject
+    public VisionSubsystem(WPIFactory factory, XPropertyManager propManager, OffboardCommunicationServer server) {
+        log.info("Creating VisionSubsystem");
```
with the new logging settings Alex did, you don't need to include the name, so this can just be `log.info("Creating");`

JohnGilb commented on Competition2017/src/competition/subsystems/vision/commands/RotateRobotToBoilerCommand.java:
```diff
@@ -0,0 +1,58 @@
+        DetectedBoiler target = visionSubsystem.getTrackedBoiler();
+        
+        rotationPid.setIMask(target == null);
+        double power = rotationPid.calculate(0, target == null ? 0 : target.offsetX);
```
This is a small thing, but I've taken to doing `rotationPid.calculate(target == null ? 0 : target.offsetX, 0);` . This means that if you need to turn left, the PID will give you a positive result, which meshes nicely with the global "positive rotation is left rotation"

WasabiFan committed:
Merge branch master

WasabiFan committed:
Add smooth drive to boiler command

WasabiFan committed:
Clean up vision code

WasabiFan committed:
Fix vision subsystem synchronization issues

WasabiFan committed:
Invert vision PID so that parameters are positive to converge

WasabiFan committed:
Merge branch 'master' into vision

WasabiFan commented:
I've updated this PR after review; this code has had a few hours of use on-robot (I was working with it until after midnight on Tuesday 😁 ) and hasn't caused any problems since I fixed that synchronization issue. The _drive to boiler with heuristics_ code is still broken, but that's an issue with the underlying drive command.

Can we aim to merge this today? I'd like to know that the base vision code is integrated so we can give it some time to soak; I will probably branch and work on some additional user-friendly optimizations now.

JohnGilb commented:
Overall still looks good - you can now take advantage of red/blue alliance detection via the pose subsystem , since you'll need two "parallel to boiler" angles (I think?)

WasabiFan committed:
Merge branch 'master' into vision
